doc/tutorials/realtime_performance_tuning.rst

```diff
@@ -0,0 +1,193 @@
+.. _rt_performance_tuning:
+
+Trace and Data Collection for ACRN Real-Time(RT) Performance Tuning
+###################################################################
+
+VM exits analysis for ACRN RT performance
+*****************************************
+
+VM exits in response to certain instructions and events are a key source of 
+performance degradation in virtual machines. During the runtime of hard RTVM 
+of ACRN, there are still some instructions and events which will impact the 
+RT latency's determinacy.
+
+  - CPUID
+  - TSC_Adjust read/write
+  - TSC write
+  - APICID/LDR read
+  - ICR write
+
+Generally, we don't want to see any VM exits occur during the critical section 
+of the RT task.
+
+The methodology of VM exits analysis is very simple. Firstly, we should clearly 
+identify the critical section of RT task. The critical section is the duration 
+of time where we do not want to see any VM exits occur. Different RT tasks get 
+different critical section. So this article will take the cyclictest as an example
+to elaborate how to do VM exits analysis.
+
+The critical sections
+=====================
+
+Here is example pseudocode of cyclictest implementation.
+
+.. code-block:: none
+
+   while (!shutdown) {
+         …
+         clock_nanosleep(&next)
+         clock_gettime(&now)
+         latency = calcdiff(now, next)
+         …
+         next += interval
+   }
+
+Time point ``now`` is the actual point at which the cyclictest is wakeuped and 
+scheduled. Time point ``next`` is the expected point at which we want the cyclictest 
+to be wakeuped and scheduled. Here we can get the latency by now - next. We don't 
+want to see VM exits during ``next`` through ``now``. So define the start point of 
+critical section as ``next`` and end point ``now``.
+
+Log and trace data collection
+=============================
+
+#. Add timestamps (in TSC) at ``next`` and ``now``.
+#. Capture the log with the above timestamps in RTVM.
+#. Capture the acrntrace log in Service VM at the same time.
+
+Offline analysis
+================
+
+#. Convert the raw trace data to human readable format.
+#. Merge the logs in RTVM and ACRN hypervisor trace based on timestamps (in TSC).
+#. Check if there are VM exits which hit the critical section, the pattern is as following:
```

Comment:
if there is any VM exit within the critical sections

Reply:
fixed.